### 13 异步 非阻塞 io

+ I/O 即 Input/Output，一个系统的输入和输出。

+ 阻塞 I/O 和非阻塞 I/O 的区别就在于
  + 系统接收输入再到输出期间，能不能接收其他输入。



### 阻塞 io 和 非阻塞的例子：

1. 食堂打饭
2. 下馆子点菜
3. 做家务做少时间，洗衣服的时候可以做其他事，



对于点菜人员：

+  排队打饭是阻塞 I/O

+  餐厅点菜是非阻塞 I/O



系统=食堂阿姨/服务生，输入=点菜，输出=端菜。

+ 饭堂阿姨只能一份一份饭地打 -> 阻塞 I/O

+ 服务生点完菜之后可以服务其他客人 -> 非阻塞 I/O







再复习一遍：

1. Node.js 是一个基于 Chrome V8 引擎的 JavaScript 运行环境。
2. Node.js 使用了一个事件驱动、非阻塞式 I/O 的模型，使其轻量又高效。



理解非阻塞 I/O 的要点在于

+ 确定一个进行 Input/Output 的系统。

+ 思考在 I/O 过程中，能不能进行其他 I/O。



通过 console.time 和 console.timeEnd 计算，中间语句消耗的时间

18.479ms

```
const glob = require('glob');

console.time('sync')
const result = glob.sync(__dirname + '/**/*')
console.timeEnd('sync')
console.log(result.length)
```

3.593ms，非阻塞

```
console.time('async')
const result2 = glob(__dirname + '/**/*', function (err, result) {
    console.log(result.length)
})
console.timeEnd('async')
// IO完成之前还可以做别的事
console.log('hello geekbang')
```



### 14 异步编程 callback

1. 回调函数格式规范

   + error-first callback

   + node-style callback

2.  第一个参数是 error，后面的参数才是结果。

```
fs.readFile('input.txt', function (err, data) {
    if (err) return console.error(err);
    console.log(data.toString());
});
```



call Stack， 调用栈

打开控制台，找到 sources ， call stack

在程序中存在，一个函数调用另一个函数的情况

每个函数调用其他函数的时候，会在调用栈中加一层，累加上去，会形成一个调用栈，最后被调用的函数在栈顶

try catch 在第五层，那么第5层以及后面的函数错误都会被捕捉到

假如，5，6，7，8，第8层，栈顶有错，会被捕获到



#### 每一个事件循环是一个新的调用栈！！！

```diff
setTimeout 函数是在另个事件循环里面
```

try catch只能抓到一个调用堆栈内，即一个事件循环里的错误

其他调用栈内的错误，不能被捕获

```js
try {
    interview(function (err, res) {
        if (err) {
            console.log('cry')
            return;
        }
        console.log('smile')
    })
} catch (e) {
    console.log('cry')
}

function interview(callback) {
    setTimeout(() => {
            throw new Error('fail');  //会被抛到全局，而不是被 try catch 捕获到的
    }, 500)
}
```

能捕获的，callback

```js
try {
    interview(function (err, res) {
        if (err) {
            console.log('cry1')
            return;
        }
        console.log('smile')
    })

} catch (e) {
    console.log('cry2')
}

function interview(callback) {
    setTimeout(() => {  
        callback(new Error('fail'))  // cry1， 满足 if(err)
    }, 500)
}

function interview(callback) {
    throw new Error('fail');  // cry2, 是在同一个事件循环里！
}
```

### 15 事件循环

思路：

当执行栈不为空时，从栈中拿出事件，然后执行

每 50毫秒执行一次 eventloop

```js
const eventloop = {
    queue: [],
    loop() {
        while(this.queue.length) {
            let callback = this.queue.shift();
            callback();
        }
        setTimeout(this.loop.bind(this), 50)
    },
    add(callback) {
        this.queue.push(callback)
    }
}

eventloop.loop();

setTimeout(()=> {
    eventloop.add(function(){ 
        console.log(1)
    })
}, 500)

setTimeout(()=> {
    eventloop.add(function(){ 
        console.log(2)
    })
}, 800)
```



```js
// 什么意思
// this绑定为 调用 setTimeout 的this
setTimeout(this.loop.bind(this), 50)
```



this 到底指向谁

```js
function LateBloomer() {
  }
  
  // 在 1 秒钟后声明 bloom
  LateBloomer.prototype.bloom = function() {
    console.log(1, this)
    setTimeout(this.declare.bind(this), 1000); // LateBloomer

    setTimeout(this.declare, 1000); //window， 相当于匿名函数

    setTimeout(()=>{
        console.log(3, this)  //LateBloomer
    })
    
    setTimeout(function(){
        console.log(4, this) // window
    })
  };
  
  LateBloomer.prototype.declare = function() {
    console.log(2, this);
  };
  
  var flower = new LateBloomer();
  flower.bloom();  // 一秒钟后, 调用 'declare' 方法
  
```

### 16 promise

问题：

then 中的 setTimeout 和 全局的 setTimeout 谁先执行？

3 0 2 4

```js
(function() {
    var promise = interview();
    var promise2 = promise.then((res)=>{
        return new Promise(function(resolve, reject) {
            setTimeout(()=>{
                console.log(2)
                resolve('acc')
            }, 400)
        })
    })


    setTimeout(()=>{
        console.log(3)
    }, 200)

    setTimeout(()=>{
        console.log(4)
    }, 1000)


})()

function interview(round) {
    return new Promise((resolve, reject) => {
        setTimeout(()=> {
            console.log(0)
            resolve(111)
        }, 500)
    })
}
```

猜顺序

3 4 2 5 1

```js
let promise = new Promise(function(resolve, reject){
    setTimeout(()=>{
        console.log(1)
        resolve(1)
    }, 5000)
})
let promise2 = new Promise(function(resolve, reject){
    console.log(3)
    resolve(1)
})
promise2.then(()=>{
    console.log(4)
    setTimeout(()=>{
        console.log(5)
    }, 1000)
})
setTimeout(()=>{
    console.log(2)
}, 0)
```



### 17. aysnc await

1.  async function 是 Promise 的语法糖封装

2. 异步编程的终极方案 – 以同步的方式写异步

   2.1 await 关键字可以“暂停”async function的执行

   2.2 await 关键字可以以同步的写法获取 Promise 的执行结果

   2.3 try-catch 可以获取 await 所得到的错误



特点1：函数会返回一个promise

```js
console.log(async function () {
}())

console.log(function(){
  return new Promise(function(resolve, reject){
    resolve(1)
  })
}()) 
```

立即执行函数的两种方式

```js
(function(){
//code
}())

(function (){
//code
})()

写在括号里 就是本文的 console.log(function(){console.log(123)}())
相当于立即执行函数
(function(){console.log(123)}())
如果单独写
function(){console.log(123)}()
不生效
```



**前面我们提到，try catch 依赖调用栈的，只能捕捉到调用栈以上的函数所抛出的错误**，异步任务这种，try catch捕获不到，但是使用 await 之后，能够被捕获到

即使这个错误在另外一个循环，另外一个函数调用栈所创建的。

```js
async function getResult(){
  try{
    var content = await new Promise((resolve, reject)=>{
        setTimeout(()=>{
            reject(new Error(8))
        }, 500)
    })
  }catch(e){
    console.log('error', e.message)
  }
  return 4
}
getResult();
```

注意，await 代码需要写在  try catch 中